### PR TITLE
feat: prometheus metrics for RPC methods

### LINKED ISSUE #3767: Add Prometheus metrics for RPC endpoints
**Issue summary**

Each RPC endpoint hit should be reflected in Forest metrics.

At the end of the day, we want to have dashboards showing how often is a given RPC method called.

Ideally, all methods would go through a single method that increases the counter on its, e.g., RPC name label.

**Other information and links**

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- Expose how many times each RPC method has been called, how long the execution took, and how many failed.

The new metrics look like this:

```
rpc_processing_time_sum{method="Filecoin.NodeStatus"} 0.006975156
rpc_processing_time_count{method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.01",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.02",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.04",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.08",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.16",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.32",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.64",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="1.28",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="2.56",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="5.12",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="+Inf",method="Filecoin.NodeStatus"} 9
```

```
rpc_method_failure{method="Filecoin.WeDontSupporThisCall"} 1
```

## Reference issue to close (if applicable)

Closes #3767

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

**Comment by LesnyRumcajs** [comment]
@lemmih is it WIP or ready to review?

File: src/metrics/mod.rs [hunk 1]
```diff
@@ -38,6 +42,29 @@ pub static LRU_CACHE_MISS: Lazy<Family<KindLabel, Counter>> = Lazy::new(|| {
...
+        "Number of failed RPC calls",
+        metric.clone(),
+    );
+    metric
+});
+
+pub static RPC_METHOD_TIME: Lazy<Family<RpcMethodLabel, Histogram>> = Lazy::new(|| {
+    let metric = Family::<RpcMethodLabel, Histogram>::new_with_constructor(|| {
+        Histogram::new(exponential_buckets(0.01, 2.0, 10))
+    });
```
Comment (LesnyRumcajs) [inline src/metrics/mod.rs:58]: What do these numbers represent?

File: src/metrics/mod.rs [hunk 2]
```diff
@@ -38,6 +42,29 @@ pub static LRU_CACHE_MISS: Lazy<Family<KindLabel, Counter>> = Lazy::new(|| {
...
+    // let metric = Family::default();
```
Comment (LesnyRumcajs) [inline src/metrics/mod.rs:59]:
```suggestion
```

**Review by LesnyRumcajs — COMMENTED** [review]
```rpc_processing_time_sum{method="Filecoin.NodeStatus"} 0.006975156
rpc_processing_time_count{method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.01",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.02",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.04",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.08",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.16",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.32",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="0.64",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="1.28",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="2.56",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="5.12",method="Filecoin.NodeStatus"} 9
rpc_processing_time_bucket{le="+Inf",method="Filecoin.NodeStatus"} 9
```
Can we avoid so many records?

File: src/metrics/mod.rs (hunk 1 above)
Comment (lemmih) [reply src/metrics/mod.rs]: Added a comment. Bucket sizes are from 10ms to 5.12s, each bucket is twice as big the previous.

**Comment by lemmih** [comment]
> ```
> rpc_processing_time_count{method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.01",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.02",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.04",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.08",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.16",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.32",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="0.64",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="1.28",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="2.56",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="5.12",method="Filecoin.NodeStatus"} 9
> rpc_processing_time_bucket{le="+Inf",method="Filecoin.NodeStatus"} 9
> ```
> 
> Can we avoid so many records?

We could track just p50, p90 and p99. But we would still have hundreds of metrics. I say we wait and see if the large number of metrics causes any problems / unexpected costs.

**Review by LesnyRumcajs — APPROVED** [review]

**Review by sudo-shashank — APPROVED** [review]

File: src/metrics/mod.rs [hunk 3]
```diff
@@ -38,6 +42,29 @@ pub static LRU_CACHE_MISS: Lazy<Family<KindLabel, Counter>> = Lazy::new(|| {
...
+    );
+    metric
+});
+
+pub static RPC_METHOD_TIME: Lazy<Family<RpcMethodLabel, Histogram>> = Lazy::new(|| {
+    let metric = Family::<RpcMethodLabel, Histogram>::new_with_constructor(|| {
+        // Histogram with 10 buckets starting from 0.01s going to 5.12s, each bucket twice as big as the last.
+        Histogram::new(exponential_buckets(0.01, 2.0, 10))
+    });
+    crate::metrics::default_registry().register(
```
Comment (qingyunha) [inline src/metrics/mod.rs:60]: Hi, I'm curious. Is there a particular reason we do not use `DEFAULT_REGISTRY.write()` here?

Comment (LesnyRumcajs) [reply src/metrics/mod.rs:60]: the `default_registry()` hides the detail that the registry is under an `RwLock`.

Comment (qingyunha) [reply src/metrics/mod.rs:60]: But, I see that all other metrics use `DEFAULT_REGISTRY.write()` in this file.

Comment (LesnyRumcajs) [reply src/metrics/mod.rs:60]: They likely shouldn't. :)

**Pull request merged**